File: 14/a.ts
// const input = await Bun.file("test.txt").text();
const input = await Bun.file("input.txt").text();

let res = 0;

const lines = input.split("\n").map((line) => line.split(""));

for (let i = 0; i < lines.length; i++) {
  for (let j = 0; j < lines[i].length; j++) {
    if (lines[i][j] !== ".") continue;
    for (let k = i + 1; k < lines.length; k++) {
      if (lines[k][j] === "#") break;
      if (lines[k][j] === "O") {
        lines[k][j] = ".";
        lines[i][j] = "O";
        break;
      }
    }
  }
}

console.log(lines.map((line) => line.join("")).join("\n"));

lines.reverse();

for (let i = 0; i < lines.length; i++) {
  for (let j = 0; j < lines[i].length; j++) {
    if (lines[i][j] === "O") res += i + 1;
  }
}

console.log(res);

File: 14/a.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	file, _ := os.ReadFile("input.txt")

	rawLines := strings.Split(string(file), "\n")
	lines := make([][]rune, len(rawLines))
	for i, line := range rawLines {
		lines[i] = []rune(line)
	}

	for j := 0; j < len(lines[0]); j++ {
		for i := 0; i < len(lines); i++ {
			if lines[i][j] != '.' {
				continue
			}
			for k := i + 1; k < len(lines); k++ {
				if lines[k][j] == '#' {
					break
				}
				if lines[k][j] == 'O' {
					lines[k][j] = '.'
					lines[i][j] = 'O'
					i++
				}
			}
		}
	}

	res := 0
	for i := 0; i < len(lines); i++ {
		for j := 0; j < len(lines[0]); j++ {
			if lines[i][j] == 'O' {
				res += len(lines) - i
			}
		}
	}

	fmt.Println(res)
}
